src/index_queue.rs

```diff
@@ -0,0 +1,273 @@
+            queue_set: CHashMap::new(),
+            map: CHashMap::new(),
+            keywords: CHashMap::new(),
+        }
+    }
+
+    pub fn enqueue(&self, item: String) {
+        if self.map.contains_key(&item) {
+            trace!("Already indexed {}", item);
+            return;
+        }
+
+        if self.queue_set.contains_key(&*item) {
+            info!("{} already in queue", item.clone());
+        } else {
+            info!("Enqueuing {}", item);
+            let _ = self.queue.push(item.clone());
+            self.queue_set.insert(item.clone(), ());
+        }
+    }
+
+    pub fn queue_length(&self) -> usize {
+        self.queue.len()
+    }
+
+    pub fn index_length(&self) -> usize {
+        self.map.len()
+    }
+
+    pub fn keyword_length(&self) -> usize {
+        self.keywords.len()
+    }
+
+    pub fn search(&self, query: String) -> Vec<IndexResult> {
+        let mut results = Vec::new();
+
+        // for the search, we could iterate through all of the indexed results and then search
+        // each result for a keyword, or instead we could do a keyword lookup in the keyword map
+        // which will give us a list of CIDs that contain the keyword.
+
+        if self.keywords.contains_key(&query) {
+            let cid_map: CHashMap<String, ()> = self.keywords.get(&query).unwrap().clone();
+            cid_map.into_iter().for_each(|(cid, _)| {
+                if self.map.contains_key(&cid) {
+                    let index_result: IndexResult = self.map.get(&cid).unwrap().clone();
+                    results.push(index_result);
+                }
+            });
+        }
+
+        return results;
+    }
+
+    pub fn start(&self, gateway: String) {
+        loop {
+            if self.queue.is_empty() == false {
+                let try_item = self.queue.pop();
+                if try_item.is_some() {
+                    let item = try_item.unwrap();
+                    self.queue_set.remove(&*item);
+                    info!("Indexing {}", item);
+
+                    let result = self.retreive_content(gateway.clone(), item.clone());
+
+                    if result.is_some() {
+                        self.map.insert(item.clone(), result.unwrap());
+                    } else {
+                        info!("Error retrieving CID {}", item);
+                        // self.enqueue(item.clone()); // for now give up on error
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Use the http client to obtain the page from the ipfs gateway. If there is a failure to
+     * obtain the CID, we give up for now.
+     */
+    fn retreive_content(&self, gateway: String, cid: String) -> Option<IndexResult> {
+        let url = format!("http://{}/ipfs/{}", gateway, cid);
+        info!("Retreiving {}", url);
+        let client = reqwest::blocking::Client::new();
+        let result = client.get(&url).send();
+        let result = match result {
```

Comment:
Minor: could just have this be `response` instead of the reassignment from `result` on the line below this pattern match block.